scripts/safe_template_linter.py

```diff
@@ -1406,6 +1419,375 @@ def _is_jquery_insert_caller_safe(self, caller_line_start):
+            self._print_node(node.func, prefix + "-")
+        elif isinstance(node, ast.Attribute):
+            print("{}Attribute: attr={}, value={}, ctx={} {}:{}".format(prefix, repr(node.attr), node.value, node.ctx, node.lineno, node.col_offset))
+            self._print_node(node.value, prefix + "-")
+        elif isinstance(node, ast.Str):
+            print("{}Str: s={} {}:{}".format(prefix, repr(node.s), node.lineno, node.col_offset))
+        elif isinstance(node, ast.Expr):
+            print("{}Expr: value={} {}:{}".format(prefix, repr(node.value), node.lineno, node.col_offset))
+        else:
+            print("{}{}".format(prefix, node))
+
+    def visit_FunctionDef(self, node):
+        """
+        Skips processing of __repr__ functions, since these sometimes use '<'
+        for non-HTML purposes.
+
+        Arguments:
+            node: An AST node.
+        """
+        if node.name != '__repr__':
+            self.generic_visit(node)
+
+
+class HtmlStringVisitor(BaseVisitor):
+    """
+    Checks for strings that contain HTML. Assumes any string with < or > is
+    considered potential HTML.
+
+    To be used only with strings in context of format or concat.
+
+    """
+    def __init__(self, file_contents, results, skip_wrapped_html=False):
+        """
+        Init function.
+
+        Arguments:
+            file_contents: The contents of the Python file.
+            results: A file results objects to which violations will be added.
+            skip_wrapped_html: True if visitor should skip strings wrapped with
+                HTML() or Text(), and False otherwise.
+        """
+        super(HtmlStringVisitor, self).__init__(file_contents, results)
+        self.skip_wrapped_html = skip_wrapped_html
+        self.unsafe_html_string_nodes = []
+        self.over_escaped_entity_string_nodes = []
+        self.has_text_or_html_call = False
+
+    def visit_Str(self, node):
+        """
+        When strings are visited, checks if it contains HTML.
+
+        Arguments:
+            node: An AST node.
+        """
+        # Skips '<' (and '>') in regex named groups. For example, "(?P<group>)".
+        if re.search('[(][?]P<', node.s) is None and re.search('[<>]', node.s) is not None:
+            self.unsafe_html_string_nodes.append(node)
+        if re.search(r"&[#]?[a-zA-Z0-9]+;", node.s):
+            self.over_escaped_entity_string_nodes.append(node)
+
+    def visit_Call(self, node):
+        """
+        Skips processing of string contained inside HTML() and Text() calls when
+        skip_wrapped_html is True.
+
+        Arguments:
+            node: An AST node.
+
+        """
+        is_html_or_text_call = isinstance(node.func, ast.Name) and node.func.id in ['HTML', 'Text']
+        if self.skip_wrapped_html and is_html_or_text_call:
+            self.has_text_or_html_call = True
+        else:
+            self.generic_visit(node)
+
+
+class CloseBeforeFormatVisitor(BaseVisitor):
```

Comment:
This visitor doesn't do anything about "close before format".  All it does is notice if there was a .format function call.